File: frontend/src/domain/models/Watchlist.ts
export type WatchSignalAction = 'BUY' | 'HOLD' | 'SELL';

export interface WatchlistSearchItem {
  symbol: string;
  name?: string;
  sector?: string;
  industry?: string;
  price?: number;
  market_cap?: number;
}

export interface I18nMessage {
  key: string;
  params?: Record<string, any>;
}

export interface WatchlistCoverageBucket {
  have: number;
  total: number;
}

export interface WatchlistCoverageBreakdown {
  framework: 'GENERAL' | 'CYCLICAL' | 'FINANCIAL' | 'UNPROFITABLE';
  fundamentals: WatchlistCoverageBucket;
  technical: WatchlistCoverageBucket;
  valuation: WatchlistCoverageBucket;
  missing_groups: string[];
}

export interface WatchlistSignal {
  action: WatchSignalAction;
  score: number;
  confidence: number;
  data_coverage: number;
  coverage_breakdown?: WatchlistCoverageBreakdown;
  freshness_days?: number;
  reasons: I18nMessage[];
}

export interface WatchlistTechnical {
  rsi14?: number;
  fifty_two_week_position?: number;
  warnings: I18nMessage[];
}

export interface WatchlistQualityProfile {
  score: number;
  outlook: 'ELITE' | 'STRONG' | 'AVERAGE' | 'SPECULATIVE';
  profitability?: number;
  growth?: number;
  financial_strength?: number;
  valuation_support?: number;
  summary: I18nMessage;
}

export interface WatchlistValueTrapRisk {
  level: 'LOW' | 'MEDIUM' | 'HIGH';
  score: number;
  reasons: I18nMessage[];
}

export interface WatchlistCycleProfile {
  is_cyclical: boolean;
  price_taker: boolean;
  earnings_regime: 'STEADY' | 'TROUGH' | 'MID' | 'PEAK';
  peak_earnings_risk: 'LOW' | 'MEDIUM' | 'HIGH';
  score: number;
  normalized_pe?: number;
  summary: I18nMessage;
  reasons: I18nMessage[];
}

export interface WatchlistTimingSignal {
  status: 'READY' | 'WAIT_PULLBACK' | 'WAIT_CONFIRMATION' | 'STALE' | 'AVOID';
  score: number;
  freshness_days?: number;
  summary: I18nMessage;
  conditions: I18nMessage[];
}

export interface WatchlistValuation {
  model: 'DCF';
  status: 'AVAILABLE' | 'UNAVAILABLE';
  valuation_method?: 'GF_BLEND_DCF_MULTIPLE';
  valuation_label?: 'DISCOUNT' | 'FAIR' | 'PREMIUM' | 'EXTREME_PREMIUM';
  fair_value?: number;
  dcf_fair_value?: number;
  dcf_upside_pct?: number;
  fair_value_low?: number;
  fair_value_high?: number;
  upside_pct?: number;
  implied_growth_10y?: number;
  discount_rate?: number;
  terminal_growth?: number;
  fcf_growth_5y?: number;
  base_fcf?: number;
  shares_outstanding?: number;
  net_debt?: number;
  risk_free_rate?: number;
  equity_risk_premium?: number;
  profitability_rank?: number;
  growth_rank?: number;
  financial_strength_rank?: number;
  valuation_rank?: number;
  gf_score?: number;
  scenarios?: {
    label: 'BEAR' | 'BASE' | 'BULL';
    fcf_growth_5y: number;
    discount_rate: number;
    fair_value: number;
  }[];
  confidence: number;
  summary: I18nMessage;
}

export type WatchPlanType = 'LONG' | 'WAIT' | 'AVOID';

export interface WatchlistTradePlan {
  plan_type: WatchPlanType;
  entry_low?: number;
  entry_high?: number;
  stop_loss?: number;
  take_profit_1?: number;
  take_profit_2?: number;
  rr_to_tp1?: number;
  rr_to_tp2?: number;
  summary: I18nMessage;
}

export interface WatchlistItem {
  symbol: string;
  name?: string;
  sector?: string;
  industry?: string;
  price?: number;
  market_cap?: number;
  forward_pe?: number;
  trailing_pe?: number;
  peg_ratio?: number;
  price_to_fcf?: number;
  revenue_growth?: number;
  eps_growth?: number;
  free_cash_flow?: number;
  roic?: number;
  roe?: number;
  updated_at?: string;
  added_at: string;
  note?: string;
  signal: WatchlistSignal;
  technical: WatchlistTechnical;
  quality: WatchlistQualityProfile;
  value_trap: WatchlistValueTrapRisk;
  cycle_profile: WatchlistCycleProfile;
  timing: WatchlistTimingSignal;
  valuation: WatchlistValuation;
  trade_plan: WatchlistTradePlan;
  // Runtime-computed industry valuation score
  valuation_score?: number;
  valuation_label?: string;
  valuation_low_confidence?: boolean;
}

export interface WatchlistResponse {
  total: number;
  items: WatchlistItem[];
}

File: backend/app/schemas/watchlist.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field


class I18nMessage(BaseModel):
    key: str
    params: dict[str, Any] = Field(default_factory=dict)


class WatchlistSearchItem(BaseModel):
    symbol: str
    name: str | None = None
    sector: str | None = None
    industry: str | None = None
    price: float | None = None
    market_cap: float | None = None


class WatchlistAddRequest(BaseModel):
    symbol: str = Field(..., min_length=1, max_length=20)
    note: str | None = Field(default=None, max_length=280)


class WatchlistUpdateRequest(BaseModel):
    note: str | None = Field(default=None, max_length=280)


class WatchlistSignal(BaseModel):
    action: Literal["BUY", "HOLD", "SELL"]
    score: float
    confidence: int
    data_coverage: float
    freshness_days: int | None = None
    reasons: list[I18nMessage] = Field(default_factory=list)


class WatchlistTechnical(BaseModel):
    rsi14: float | None = None
    fifty_two_week_position: float | None = None
    warnings: list[I18nMessage] = Field(default_factory=list)


class WatchlistValuationScenario(BaseModel):
    label: Literal["BEAR", "BASE", "BULL"]
    fcf_growth_5y: float
    discount_rate: float
    fair_value: float


class WatchlistValuation(BaseModel):
    model: Literal["DCF"] = "DCF"
    status: Literal["AVAILABLE", "UNAVAILABLE"]
    valuation_method: Literal["GF_BLEND_DCF_MULTIPLE"] | None = None
    valuation_label: Literal["DISCOUNT", "FAIR", "PREMIUM", "EXTREME_PREMIUM"] | None = None
    fair_value: float | None = None
    dcf_fair_value: float | None = None
    dcf_upside_pct: float | None = None
    fair_value_low: float | None = None
    fair_value_high: float | None = None
    upside_pct: float | None = None
    implied_growth_10y: float | None = None
    discount_rate: float | None = None
    terminal_growth: float | None = None
    fcf_growth_5y: float | None = None
    base_fcf: float | None = None
    shares_outstanding: float | None = None
    net_debt: float | None = None
    risk_free_rate: float | None = None
    equity_risk_premium: float | None = None
    profitability_rank: int | None = None
    growth_rank: int | None = None
    financial_strength_rank: int | None = None
    valuation_rank: int | None = None
    gf_score: int | None = None
    scenarios: list[WatchlistValuationScenario] = Field(default_factory=list)
    confidence: int
    summary: I18nMessage


class WatchlistTradePlan(BaseModel):
    plan_type: Literal["LONG", "WAIT", "AVOID"]
    entry_low: float | None = None
    entry_high: float | None = None
    stop_loss: float | None = None
    take_profit_1: float | None = None
    take_profit_2: float | None = None
    rr_to_tp1: float | None = None
    rr_to_tp2: float | None = None
    summary: I18nMessage


class WatchlistDCFSimulationRequest(BaseModel):
    symbol: str
    price: float
    base_value: float  # EPS or FCF
    discount_rate: float
    growth_rate: float
    growth_years: int = 10
    terminal_growth: float = 0.04
    terminal_years: int = 10
    tangible_book_value: float = 0.0
    add_tangible_book: bool = False


class WatchlistDCFSimulationResponse(BaseModel):
    fair_value: float
    growth_value: float
    terminal_value: float
    margin_of_safety: float
    implied_growth: float | None = None


class WatchlistItem(BaseModel):
    symbol: str
    name: str | None = None
    sector: str | None = None
    industry: str | None = None
    price: float | None = None
    market_cap: float | None = None
    forward_pe: float | None = None
    trailing_pe: float | None = None
    peg_ratio: float | None = None
    price_to_fcf: float | None = None
    revenue_growth: float | None = None
    eps_growth: float | None = None
    free_cash_flow: float | None = None
    roic: float | None = None
    roe: float | None = None
    updated_at: datetime | None = None
    added_at: datetime
    note: str | None = None
    signal: WatchlistSignal
    technical: WatchlistTechnical
    valuation: WatchlistValuation
    trade_plan: WatchlistTradePlan
    # Runtime-computed — not stored in DB
    valuation_score: float | None = None
    valuation_label: str | None = None
    valuation_low_confidence: bool | None = None


class WatchlistResponse(BaseModel):
    total: int
    items: list[WatchlistItem]
